piecedirectory/doctor.go

```diff
@@ -5,38 +5,65 @@ import (
 	"errors"
 	"fmt"
 	"math/rand"
+	"sync"
 	"time"
 
-	"github.com/filecoin-project/boost/piecedirectory/types"
+	"github.com/filecoin-project/boost/db"
+	"github.com/filecoin-project/boost/sectorstatemgr"
+	bdclient "github.com/filecoin-project/boostd-data/client"
+	"github.com/filecoin-project/go-address"
 	"github.com/filecoin-project/go-state-types/abi"
 	"github.com/ipfs/go-cid"
 	logging "github.com/ipfs/go-log/v2"
 )
 
 var doclog = logging.Logger("piecedoc")
 
-type SealingApi interface {
-	IsUnsealed(ctx context.Context, sectorID abi.SectorNumber, offset abi.UnpaddedPieceSize, length abi.UnpaddedPieceSize) (bool, error)
-}
-
 // The Doctor periodically queries the local index directory for piece cids, and runs
 // checks against those pieces. If there is a problem with a piece, it is
 // flagged, so that it can be surfaced to the user.
 // Note that multiple Doctor processes can run in parallel. The logic for which
 // pieces to give to the Doctor to check is in the local index directory.
 type Doctor struct {
-	store types.Store
-	sapi  SealingApi
+	store *bdclient.Store
+	ssm   *sectorstatemgr.SectorStateMgr
+
+	latestUpdateMu sync.Mutex
+	latestUpdate   *sectorstatemgr.SectorStateUpdates
 }
 
-func NewDoctor(store types.Store, sapi SealingApi) *Doctor {
-	return &Doctor{store: store, sapi: sapi}
+func NewDoctor(store *bdclient.Store, ssm *sectorstatemgr.SectorStateMgr) *Doctor {
+	return &Doctor{store: store, ssm: ssm}
 }
 
 // The average interval between calls to NextPiecesToCheck
 const avgCheckInterval = 30 * time.Second
 
 func (d *Doctor) Run(ctx context.Context) {
+	doclog.Info("piece doctor: running")
+
+	go func() {
+		sub := d.ssm.PubSub.Subscribe()
+
+		for {
+			select {
+			case u, ok := <-sub:
+				if !ok {
+					log.Debugw("state updates subscription closed")
+					return
+				}
+				log.Debugw("got state updates from SectorStateMgr", "len(u.updates)", len(u.Updates), "len(u.active)", len(u.ActiveSectors), "u.updatedAt", u.UpdatedAt)
+
+				d.latestUpdateMu.Lock()
+				d.latestUpdate = u
+				d.latestUpdateMu.Unlock()
+
```

Comment:
Does it make sense to trigger a check in response to getting a new update from SectorStateMgr or are they on very different schedules?